Lazy accessor synthesis for storage in primary files

slavapestov opened the pull request:
This is a follow-up to https://github.com/apple/swift/pull/26461. It introduces new methods on AbstractStorageDecl for visiting accessors in a deterministic way, intended to replace getAllAccessors():

- visitParsedAccessors() - visits parsed accessors only
- visitEmittedAccessors() - visits parsed and opaque accessors

This facilitates the removal of `addExpectedOpaqueAccessorsToStorage()`.

slavapestov committed:
Sema: Remove addExpectedOpaqueAccessorsToStorage()

slavapestov committed:
AST: Replace AbstractStorageDecl::hasAnyAccessors() with hasParsedAccessors()

slavapestov committed:
AST: Introduce AbstractStorageDecl::visit{Parsed,Emitted}Accessors()

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov committed:
AST: Slightly more efficient requiresNewVTableEntry()

rjmccall commented on include/swift/AST/Decl.h:
```diff
@@ -4609,13 +4608,19 @@ class AbstractStorageDecl : public ValueDecl {
   void visitOpaqueAccessors(llvm::function_ref<void (AccessorDecl*)>) const;
 
+  /// Visit all parsed and opaque accessors.
+  void visitEmittedAccessors(llvm::function_ref<void (AccessorDecl*)>) const;
```
"Eagerly" emitted accessors, maybe?

rjmccall commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2825,10 +2826,12 @@ void TypeChecker::addImplicitDynamicAttribute(Decl *D) {
+    bool blocked = false;
+    asd->visitParsedAccessors([&](AccessorDecl *accessor) {
+      blocked |= shouldBlockImplicitDynamic(accessor);
+    });
```
Please fix the typo in the comment while you're here.

rjmccall commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2542,9 +2538,8 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
-      // addExpectedOpaqueAccessorsToStorage(), to ensure they run when a
-      // protocol requirement is validated but not type checked.
+      // Stored properties in protocols are diagnosed in elsewhere; see
+      // StorageImplInfoRequest::evaluate().
```
"in elsewhere"

rjmccall reviewed (commented):
Minor requests only

slavapestov committed:
AST: Fix mangling of unvalidated declarations

This is again just a temporary workaround until getInterfaceType() can
kick off a request.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please smoke test source compatibility

slavapestov commented:
@swift-ci Please smoke test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility debug

slavapestov commented:
@swift-ci Please test source compatibility debug

slavapestov commented:
@swift-ci Please test source compatibility debug
